src/qibo/backends/numpy.py

```diff
@@ -702,9 +715,9 @@ def calculate_norm_density_matrix(self, state, order="nuc"):
         return self.np.linalg.norm(state, ord=order)
 
     def calculate_overlap(self, state1, state2):
-        state1 = self.cast(state1)
-        state2 = self.cast(state2)
-        return self.np.abs(self.np.sum(self.np.conj(state1) * state2))
+        return self.np.abs(
+            self.np.sum(self.np.conj(self.cast(state1)) * self.cast(state2))
+        )
```

Comment:
Not related to this PR per se, but this method should be removed entirely since there are already functions that do the same thing in `quantum_info`